### ISSUE TITLE
2.7.11,java.lang.NoSuchMethodError: org.apache.curator.framework.recipes.cache.NodeCache.getListenable

### ISSUE THREAD

**Issue by NickZhang1024** [apache/dubbo#7772]
- [x] I have searched the [issues](https://github.com/apache/dubbo/issues) of this repository and believe that this is not a duplicate.
- [x] I have checked the [FAQ](https://github.com/apache/dubbo/blob/master/FAQ.md) of this repository and believe that this is not a duplicate.

### Environment

* Dubbo version: 2.7.11
* Operating System version: Windows 10
* Java version: 1.8

### Steps to reproduce this issue

1. pom.xml
        
        <dependency>
            <groupId>org.apache.dubbo</groupId>
            <artifactId>dubbo-spring-boot-starter</artifactId>
        </dependency>
        <dependency>
            <groupId>org.apache.dubbo</groupId>
            <artifactId>dubbo</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>org.javassist</groupId>
                    <artifactId>javassist</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.jboss.netty</groupId>
                    <artifactId>netty</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.springframework</groupId>
                    <artifactId>spring-context</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        
        <dependency>
            <groupId>org.apache.curator</groupId>
            <artifactId>curator-framework</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>com.google.guava</groupId>
                    <artifactId>guava</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.apache.zookeeper</groupId>
                    <artifactId>zookeeper</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>org.apache.curator</groupId>
            <artifactId>curator-recipes</artifactId>
        </dependency>
        
        <dependency>
            <groupId>org.apache.zookeeper</groupId>
            <artifactId>zookeeper</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>jline</groupId>
                    <artifactId>jline</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>io.netty</groupId>
                    <artifactId>netty</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>io.netty</groupId>
                    <artifactId>netty-handler</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>io.netty</groupId>
                    <artifactId>netty-transport-native-epoll</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>log4j</groupId>
                    <artifactId>log4j</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.slf4j</groupId>
                    <artifactId>slf4j-api</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.slf4j</groupId>
                    <artifactId>slf4j-log4j12</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.apache.yetus</groupId>
                    <artifactId>audience-annotations</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>org.apache.curator</groupId>
            <artifactId>curator-x-discovery</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>org.apache.zookeeper</groupId>
                    <artifactId>zookeeper</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        
        <dependency>
            <groupId>io.netty</groupId>
            <artifactId>netty-all</artifactId>
        </dependency>

Pls. provide [GitHub address] to reproduce this issue.

### Result

dubbo 2.7.10 good running, but update dubbo 2.7.11 run error,

 java.lang.NoSuchMethodError: org.apache.curator.framework.recipes.cache.NodeCache.getListenable()Lorg/apache/curator/framework/listen/ListenerContainer;
java.lang.IllegalStateException: Extension instance (name: registry, class: interface org.apache.dubbo.rpc.Protocol) couldn't be instantiated: null

![image](https://user-images.githubusercontent.com/834034/118391823-d7e69900-b668-11eb-878a-4096ad5dabb3.png)

![image](https://user-images.githubusercontent.com/834034/118391763-9b1aa200-b668-11eb-9acb-67052f4cfed5.png)

**Comment by icankeep** [comment]
add zookeeper dependency like this would be better
```xml
  <dependency>
      <groupId>org.apache.dubbo</groupId>
      <artifactId>dubbo-dependencies-zookeeper</artifactId>
      <version>${dubbo.version}</version>
      <type>pom</type>
      <scope>import</scope>
  </dependency>
```

**Comment by NickZhang1024** [comment]
still error @icankeep 

[WARN ] 11:57:47.577 (               org.apache.dubbo.common.config.ConfigurationUtils) [                    main] -  [DUBBO] You specified the config center, but there's not even one single config item in it., dubbo version: 2.7.11, current host: 192.168.233.1
[WARN ] 11:57:47.577 (               org.apache.dubbo.common.config.ConfigurationUtils) [                    main] -  [DUBBO] You specified the config center, but there's not even one single config item in it., dubbo version: 2.7.11, current host: 192.168.233.1
[ERROR] 11:57:47.711 (.springframework.boot.diagnostics.LoggingFailureAnalysisReporter) [                    main] - 

***************************
APPLICATION FAILED TO START
***************************

Description:

An attempt was made to call a method that does not exist. The attempt was made from the following location:

    org.apache.dubbo.remoting.zookeeper.curator.CuratorZookeeperClient.addTargetDataListener(CuratorZookeeperClient.java:239)

The following method did not exist:

    org.apache.curator.framework.recipes.cache.NodeCache.getListenable()Lorg/apache/curator/framework/listen/ListenerContainer;

The method's class, org.apache.curator.framework.recipes.cache.NodeCache, is available from the following locations:

    jar:file:/D:/Repositories/Maven/org/apache/curator/curator-recipes/5.1.0/curator-recipes-5.1.0.jar!/org/apache/curator/framework/recipes/cache/NodeCache.class

The class hierarchy was loaded from the following locations:

    org.apache.curator.framework.recipes.cache.NodeCache: file:/D:/Repositories/Maven/org/apache/curator/curator-recipes/5.1.0/curator-recipes-5.1.0.jar


Action:

Correct the classpath of your application so that it contains a single, compatible version of org.apache.curator.framework.recipes.cache.NodeCache

java.lang.IllegalStateException: Extension instance (name: registry, class: interface org.apache.dubbo.rpc.Protocol) couldn't be instantiated: null
	at org.apache.dubbo.common.extension.ExtensionLoader.createExtension(ExtensionLoader.java:672)
	at org.apache.dubbo.common.extension.ExtensionLoader.getExtension(ExtensionLoader.java:437)
	at org.apache.dubbo.common.extension.ExtensionLoader.getExtension(ExtensionLoader.java:421)
	at org.apache.dubbo.rpc.Protocol$Adaptive.export(Protocol$Adaptive.java)
	at org.apache.dubbo.config.ServiceConfig.doExportUrlsFor1Protocol(ServiceConfig.java:513)
	at org.apache.dubbo.config.ServiceConfig.doExportUrls(ServiceConfig.java:338)
	at org.apache.dubbo.config.ServiceConfig.doExport(ServiceConfig.java:314)
	at org.apache.dubbo.config.ServiceConfig.export(ServiceConfig.java:208)
	at org.apache.dubbo.config.bootstrap.DubboBootstrap.exportService(DubboBootstrap.java:1081)
	at org.apache.dubbo.config.bootstrap.DubboBootstrap.lambda$exportServices$14(DubboBootstrap.java:1068)
	at java.util.HashMap$Values.forEach(HashMap.java:981)
	at org.apache.dubbo.config.bootstrap.DubboBootstrap.exportServices(DubboBootstrap.java:1052)
	at org.apache.dubbo.config.bootstrap.DubboBootstrap.start(DubboBootstrap.java:877)
	at org.apache.dubbo.config.spring.context.DubboBootstrapApplicationListener.onContextRefreshedEvent(DubboBootstrapApplicationListener.java:69)
	at org.apache.dubbo.config.spring.context.DubboBootstrapApplicationListener.onApplicationContextEvent(DubboBootstrapApplicationListener.java:62)
	at com.alibaba.spring.context.OnceApplicationContextEventListener.onApplicationEvent(OnceApplicationContextEventListener.java:52)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:176)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:169)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:143)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:421)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:378)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:938)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:586)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:782)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:774)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:439)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:339)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1340)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1329)
	at com.iknowpower.bm.etsms.evcar.ccs.springboot.BmEtsmsEvcarCcsSpringBootApplication.main(BmEtsmsEvcarCcsSpringBootApplication.java:19)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.dubbo.common.extension.ExtensionLoader.createExtension(ExtensionLoader.java:642)
	... 30 more
Caused by: java.lang.NoSuchMethodError: org.apache.curator.framework.recipes.cache.NodeCache.getListenable()Lorg/apache/curator/framework/listen/ListenerContainer;
	at org.apache.dubbo.remoting.zookeeper.curator.CuratorZookeeperClient.addTargetDataListener(CuratorZookeeperClient.java:239)
	at org.apache.dubbo.remoting.zookeeper.curator.CuratorZookeeperClient.addTargetDataListener(CuratorZookeeperClient.java:53)
	at org.apache.dubbo.remoting.zookeeper.support.AbstractZookeeperClient.addDataListener(AbstractZookeeperClient.java:124)
	at org.apache.dubbo.configcenter.support.zookeeper.ZookeeperDynamicConfiguration.doAddListener(ZookeeperDynamicConfiguration.java:111)
	at org.apache.dubbo.common.config.configcenter.TreePathDynamicConfiguration.addListener(TreePathDynamicConfiguration.java:102)
	at org.apache.dubbo.common.config.configcenter.wrapper.CompositeDynamicConfiguration.lambda$addListener$0(CompositeDynamicConfiguration.java:45)
	at org.apache.dubbo.common.config.configcenter.wrapper.CompositeDynamicConfiguration.iterateListenerOperation(CompositeDynamicConfiguration.java:87)
	at org.apache.dubbo.common.config.configcenter.wrapper.CompositeDynamicConfiguration.addListener(CompositeDynamicConfiguration.java:45)
	at org.apache.dubbo.rpc.cluster.governance.DefaultGovernanceRuleRepositoryImpl.addListener(DefaultGovernanceRuleRepositoryImpl.java:28)
	at org.apache.dubbo.rpc.cluster.governance.GovernanceRuleRepository.addListener(GovernanceRuleRepository.java:34)
	at org.apache.dubbo.registry.integration.AbstractConfiguratorListener.initWith(AbstractConfiguratorListener.java:45)
	at org.apache.dubbo.registry.integration.RegistryProtocol$ProviderConfigurationListener.<init>(RegistryProtocol.java:736)
	at org.apache.dubbo.registry.integration.RegistryProtocol.<init>(RegistryProtocol.java:137)
	at org.apache.dubbo.registry.integration.InterfaceCompatibleRegistryProtocol.<init>(InterfaceCompatibleRegistryProtocol.java:38)
	... 35 more

**Comment by icankeep** [comment]
are you sure remove curator dependencies and add dubbo-dependencies-zookeeper in your pom file?
![image](https://user-images.githubusercontent.com/39434627/118431622-5a269a00-b709-11eb-9c42-12f0c52b3d31.png)

dubbo 2.7.11 -->  
```xml
<zookeeper_version>3.4.13</zookeeper_version>
<curator_version>4.0.1</curator_version>
```

**Comment by icankeep** [comment]
put your pom.xml with pretty fomat wil be better

**Comment by NickZhang1024** [comment]
but 2.7.10 is running good

**Comment by NickZhang1024** [comment]
because project include elasticjob 3.0.0-RC1, it must include curator 5.1.0. but dubbo 2.7.10 use curator 5.1.0  is running good

**Comment by MentosL** [comment]
@NickZhang1024  Can provide a demo project？

**Comment by icankeep** [comment]
it isn't a problem in dubbo 2.7.11, there are some conflits in your dependencies


you can modify your pom file like the following demo
[demo](https://github.com/icankeep/dubbo-samples/tree/debug%23dubbo-issue-7772/dubbo-samples-zookeeper)

I can run normally in this demo

```xml
<dependency>
    <groupId>org.apache.curator</groupId>
    <artifactId>curator-recipes</artifactId>
    <version>5.1.0</version>
    <exclusions>
        <exclusion>
            <artifactId>curator-framework</artifactId>
            <groupId>org.apache.curator</groupId>
        </exclusion>
    </exclusions>
</dependency>
<!-- https://mvnrepository.com/artifact/org.apache.curator/curator-framework -->
<dependency>
    <groupId>org.apache.curator</groupId>
    <artifactId>curator-framework</artifactId>
    <version>5.1.0</version>
    <exclusions>
        <exclusion>
            <artifactId>guava</artifactId>
            <groupId>com.google.guava</groupId>
        </exclusion>
    </exclusions>
</dependency>
<!-- https://mvnrepository.com/artifact/org.apache.curator/curator-x-discovery -->
<dependency>
    <groupId>org.apache.curator</groupId>
    <artifactId>curator-x-discovery</artifactId>
    <version>5.1.0</version>
    <exclusions>
        <exclusion>
            <artifactId>curator-recipes</artifactId>
            <groupId>org.apache.curator</groupId>
        </exclusion>
    </exclusions>
</dependency>
<!-- https://mvnrepository.com/artifact/org.apache.zookeeper/zookeeper -->
<dependency>
    <groupId>org.apache.zookeeper</groupId>
    <artifactId>zookeeper</artifactId>
    <version>3.7.0</version>
</dependency>
<!-- https://mvnrepository.com/artifact/org.apache.shardingsphere.elasticjob/elasticjob-api -->
<dependency>
    <groupId>org.apache.shardingsphere.elasticjob</groupId>
    <artifactId>elasticjob-api</artifactId>
    <version>3.0.0-RC1</version>
</dependency>
<!-- https://mvnrepository.com/artifact/org.apache.shardingsphere.elasticjob/elasticjob-lite-core -->
<dependency>
    <groupId>org.apache.shardingsphere.elasticjob</groupId>
    <artifactId>elasticjob-lite-core</artifactId>
    <version>3.0.0-RC1</version>
    <exclusions>
        <exclusion>
            <artifactId>guava</artifactId>
            <groupId>com.google.guava</groupId>
        </exclusion>
        <exclusion>
            <artifactId>curator-client</artifactId>
            <groupId>org.apache.curator</groupId>
        </exclusion>
        <exclusion>
            <artifactId>curator-framework</artifactId>
            <groupId>org.apache.curator</groupId>
        </exclusion>
        <exclusion>
            <artifactId>curator-recipes</artifactId>
            <groupId>org.apache.curator</groupId>
        </exclusion>
    </exclusions>
</dependency>
```

**Comment by harvies** [comment]
Version 2.7.11 changed the value of nodeCacheMap from the original TreeCache to NodeCache. The return value of the getListenable method of NodeCache in the cursor-recipes4.0.1 that dubbo relies on is ListenerContainer type, and the return value of the getListenable method of NodeCache in cursor-recipes 5.0.1 is Listenable type.
![image](https://user-images.githubusercontent.com/16585330/124935423-3c562100-e038-11eb-87c3-c31fad5b634f.png)

![image](https://user-images.githubusercontent.com/16585330/124934998-d23d7c00-e037-11eb-9530-6b1f6bd2805d.png)
![image](https://user-images.githubusercontent.com/16585330/124935045-dcf81100-e037-11eb-93bb-a7bac6290a19.png)
![image](https://user-images.githubusercontent.com/16585330/124935226-0d3faf80-e038-11eb-9cdc-554a602c13d8.png)
![image](https://user-images.githubusercontent.com/16585330/124935354-28122400-e038-11eb-8eed-2eb87688a965.png)

**Comment by horizonzy** [comment]
Hi, it seems that curator's api compatible problem.  NodeCache in curator 5.x.0 getListenable() return Listenable.

**Comment by horizonzy** [comment]
Hi, it seems that curator's api compatible problem.  
NodeCache in curator 5.x getListenable() return Listenable.
NodeCache in curator 4.x getListenable() return  ListenerContainer.
If user import curator 5.x in there project, it will cover dubbo's curator version. So cause the compatible problem.

**Comment by harvies** [comment]
> Hi, it seems that curator's api compatible problem. We use curator 4.x in dubbo.
> NodeCache in curator 5.x getListenable() return Listenable.
> NodeCache in curator 4.x getListenable() return ListenerContainer.
> If user import curator 5.x in there project, it will cover dubbo's curator version. So cause the compatible problem.

Version 2.7.10 is compatible with curator 5.x, and upgrading to 2.7.11 is not compatible.

**Comment by horizonzy** [comment]
> Version 2.7.10 is compatible with curator 5.x, and upgrading to 2.7.11 is not compatible.
Yes, we use nodeCache to replace TreeCache at 2.7.11, but NodeCache is not compatible with 5.x and 4.x. So the problem occur.
We will supply a way to solve this problem.

**Comment by horizonzy** [comment]
@harvies Hi, harvies. I pull request a patch #8264 for this issue. Can you spare time to help test it. I have test it, it works fine.

**Comment by harvies** [comment]
> @harvies Hi, harvies. I pull request a patch #8264 for this issue. Can you spare time to help test it. I have test it, it works fine.

Thank you, I will test it later.

**Comment by harvies** [comment]
> @harvies Hi, harvies. I pull request a patch #8264 for this issue. Can you spare time to help test it. I have test it, it works fine.

I tested it and it works fine.

**Comment by horizonzy** [comment]
> I tested it and it works fine.

Thanks.

**Comment by CrazyHZM** [comment]
Resolved in #8264
